astacus/coordinator/plugins/clickhouse/async_object_storage.py

```diff
@@ -46,30 +57,70 @@ def delete_key(self, key: str) -> None:
         with self._storage_lock:
             self._storage.delete_key(key)
 
+    def copy_items_from(self, source: "ThreadSafeRohmuStorage", keys: Sequence[str]) -> None:
+        # In theory this could deadlock if some other place was locking the same two storages
+        # in the reverse order at the same time. Within the context of backups and restore,
+        # it's quite unlikely to have a pair of storages be the source and target of each other.
+        # Especially since we create new storage objects for each coordinator operation.
+        with source.get_storage() as source_storage:
+            with self.get_storage() as target_storage:
+                target_storage.copy_files_from(source=source_storage, keys=keys)
+
+    @contextlib.contextmanager
+    def get_storage(self) -> Iterator[BaseTransfer]:
+        with self._storage_lock:
+            yield self._storage
+
 
 @dataclasses.dataclass(frozen=True)
 class RohmuAsyncObjectStorage(AsyncObjectStorage):
     storage: ThreadSafeRohmuStorage
 
+    def get_config(self) -> Mapping[str, Any]:
+        return self.storage.config
+
     async def list_items(self) -> list[ObjectStorageItem]:
         items = await run_in_threadpool(self.storage.list_iter, key="", with_metadata=True, deep=True)
         return [ObjectStorageItem(key=Path(item["name"]), last_modified=item["last_modified"]) for item in items]
 
     async def delete_item(self, key: Path) -> None:
         await run_in_threadpool(self.storage.delete_key, str(key))
 
+    async def copy_items_from(self, source: "AsyncObjectStorage", keys: Sequence[Path]) -> None:
+        if not isinstance(source, RohmuAsyncObjectStorage):
+            raise NotImplementedError("Copying items is only supported from another RohmuAsyncObjectStorage")
+        str_keys = [str(key) for key in keys]
+        await run_in_threadpool(self.storage.copy_items_from, source.storage, str_keys)
```

Comment:
What advantage is there in using `starlette`'s `run_in_threadpool` over stdlib's `loop.run_in_executor`?

Reply:
I don't think it has a big impact for us but [starlette](https://github.com/encode/starlette/blob/7e6c936487bc6bc57f15679fcadb8b9589d43636/starlette/concurrency.py#L29-L35) uses anyio to use different implementations but we only use the vanilla asyncio & thread.
Though, they take care of contextvars, that's one less thing to worry about: https://github.com/agronholm/anyio/blob/a9fae31ccfef0803bf14ab65bd1e9b76c7fc4920/src/anyio/_backends/_asyncio.py#L2055-L2106

It's there mostly for consistency since we use the same function in other places of astacus.